webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/facts/nodes/NodeFact.scala

```diff
@@ -1289,17 +1302,139 @@ object NodeFactChangeEvent {
   }
 }
 
+/*
+ * A change context groups together information needed to track a change: who, when, why
+ */
 final case class ChangeContext(
     modId:     ModificationId,
     actor:     EventActor,
     eventDate: DateTime,
     message:   Option[String],
-    actorIp:   Option[String]
+    actorIp:   Option[String],
+    nodePerms: NodeSecurityContext
 )
 
 object ChangeContext {
-  def newForRudder(msg: Option[String] = None, actorIp: Option[String] = None) =
-    ChangeContext(ModificationId(java.util.UUID.randomUUID.toString), eventlog.RudderEventActor, DateTime.now(), msg, actorIp)
+  implicit class ChangeContextImpl(cc: ChangeContext) {
+    def toQuery: QueryContext = QueryContext(cc.actor, cc.nodePerms)
+  }
+
+  def newForRudder(msg: Option[String] = None, actorIp: Option[String] = None) = {
+    ChangeContext(
+      ModificationId(java.util.UUID.randomUUID.toString),
+      eventlog.RudderEventActor,
+      DateTime.now(),
+      msg,
+      actorIp,
+      NodeSecurityContext.All
+    )
+  }
+}
+
+/*
+ * A query context groups together information that are needed to either filter out
+ * some result regarding a security context, or to enhance query efficiency by limiting
+ * the item to retrieve. It's granularity is at the item level, not attribute level. For that
+ * latter need, by-item solution need to be used (see for ex: SelectFacts for nodes)
+ */
+final case class QueryContext(
+    actor:     EventActor,
+    nodePerms: NodeSecurityContext
+)
+
+object QueryContext {
+  // for test
+  implicit val testQC: QueryContext = QueryContext(eventlog.RudderEventActor, NodeSecurityContext.All)
```

Comment:
I think it should be created in test classes. to prevent we use it in code by mistake

Reply:
but that could be done in another time i think